File: gmm/scala/gmm.scala
import org.apache.spark.ml.clustering.{GaussianMixture, GaussianMixtureModel}
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.sql.SparkSession

object GMMExample {
  def main(args: Array[String]): Unit = {
    val spark = SparkSession.builder()
      .appName("GMMExample")
      .master("local[*]")
      .getOrCreate()

    val NSamples = 500
    val NComponents = 2

    // Generate data from the first Gaussian distribution
    val data1 = Array.fill(NSamples / 2)(Vectors.dense(util.Random.nextGaussian(), util.Random.nextGaussian()))

    // Generate data from the second Gaussian distribution
    val data2 = Array.fill(NSamples / 2)(Vectors.dense(util.Random.nextGaussian() + 3, util.Random.nextGaussian() + 3))

    // Combine the data from both distributions
    val data = spark.createDataFrame(data1 ++ data2).toDF("features")

    // Fit the Gaussian Mixture Model
    val gmm = new GaussianMixture()
      .setK(NComponents)
      .fit(data)

    // Retrieve the GMM parameters
    val weights = gmm.weights
    val means = gmm.gaussians.map(_.mean)
    val covariances = gmm.gaussians.map(_.cov)

    // Print the results
    println("Weights:")
    weights.foreach(w => print(w + " "))

    println("\n\nMeans:")
    means.foreach(m => println(m.toArray.mkString(" ")))

    println("\nCovariances:")
    covariances.foreach(c => println(c.toArray.mkString(" ")))

    spark.stop()
  }
}

File: gmm/matlab/gmm.m
rng('default');  % Set the random number generator seed for reproducibility

NSamples = 500;
NComponents = 2;

% Generate data from the first Gaussian distribution
data1 = mvnrnd([0 0], [1 0; 0 1], NSamples/2);

% Generate data from the second Gaussian distribution
data2 = mvnrnd([3 3], [1 0; 0 1], NSamples/2);

% Combine the data from both distributions
data = [data1; data2];

% Fit the Gaussian Mixture Model
gmm = fitgmdist(data, NComponents);

% Retrieve the GMM parameters
weights = gmm.ComponentProportion;
means = gmm.mu;
covariances = gmm.Sigma;

% Print the results
disp('Weights:');
disp(weights);

disp('Means:');
disp(means);

disp('Covariances:');
for i = 1:NComponents
    disp(covariances(:, :, i));
end
